dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/taskgroup/TaskGroupCoordinator.java

```diff
@@ -0,0 +1,451 @@
+            } catch (Throwable throwable) {
+                log.info("Notify the force start TaskGroupQueue {} failed", taskGroupQueue, throwable);
+            } finally {
+                LogUtils.removeTaskInstanceIdMDC();
+            }
+        }
+    }
+
+    private void dealWithWaitingTaskGroupQueue() {
+        // Find the TaskGroup which usage < maxSize.
+        // Find the highest priority inQueue task group queue(Which is inQueue and status is Waiting and force start is
+        // NO) belong to the
+        // task group.
+        List<TaskGroup> taskGroups = taskGroupDao.queryAvailableTaskGroups();
+        if (CollectionUtils.isEmpty(taskGroups)) {
+            log.debug("There is no available task group");
+            return;
+        }
+        for (TaskGroup taskGroup : taskGroups) {
+            int availableSize = taskGroup.getGroupSize() - taskGroup.getUseSize();
+            if (availableSize <= 0) {
+                log.info("TaskGroup {} is full, available size is {}", taskGroup, availableSize);
+                continue;
+            }
+            List<TaskGroupQueue> taskGroupQueues =
+                    taskGroupQueueDao.queryAllInQueueTaskGroupQueueByGroupId(taskGroup.getId())
+                            .stream()
+                            .filter(taskGroupQueue -> Flag.NO.getCode() == taskGroupQueue.getForceStart())
+                            .filter(taskGroupQueue -> TaskGroupQueueStatus.WAIT_QUEUE == taskGroupQueue.getStatus())
+                            .limit(availableSize)
+                            .collect(Collectors.toList());
+            if (CollectionUtils.isEmpty(taskGroupQueues)) {
+                log.debug("There is no waiting task group queue for task group {}", taskGroup.getName());
+                continue;
+            }
+            for (TaskGroupQueue taskGroupQueue : taskGroupQueues) {
+                try {
+                    LogUtils.setTaskInstanceIdMDC(taskGroupQueue.getTaskId());
+                    // Reduce the taskGroupSize
+                    boolean acquireResult = taskGroupDao.acquireTaskGroupSlot(taskGroup.getId());
+                    if (!acquireResult) {
+                        log.error("Failed to acquire task group slot for task group {}", taskGroup);
+                        continue;
+                    }
+                    // Notify the waiting task instance
+                    // We notify first, it notify failed, the taskGroupQueue will be in queue, and then we will retry it
+                    // next time.
+                    notifyWaitingTaskInstance(taskGroupQueue);
+
+                    // Set the taskGroupQueue status to RUNNING and remove from queue
+                    taskGroupQueue.setInQueue(Flag.YES.getCode());
+                    taskGroupQueue.setStatus(TaskGroupQueueStatus.ACQUIRE_SUCCESS);
+                    taskGroupQueue.setUpdateTime(new Date());
+                    taskGroupQueueDao.updateById(taskGroupQueue);
```

Comment:
There is some repetitive code in the state change that might be suitable for turning into a function.

Reply:
Good suggestion, I will use another PR to move these kind of code in the whole project into function.